File: GPS/zone.py
import rospy
from sensor_msgs.msg import NavSatFix
from geometry_msgs.msg import PointStamped
from math import sqrt

class Zone:
    def __init__(self):
        self.global_latitude = 0.0
        self.global_longitude = 0.0
        self.P1 = None
        self.P2 = None
        self.P3 = None
        self.P4 = None

    def gps_callback(self, msg):
        # Callback appelée lors de la réception de messages GPS
        self.global_latitude = msg.latitude
        self.global_longitude = msg.longitude
        rospy.loginfo("Coordonnées GPS : Latitude = %f, Longitude = %f", self.global_latitude, self.global_longitude)

    def is_inside_rectangle(self, A, B, C, D, P):
        # Vérifier si le point est à l'intérieur du rectangle ABCD
        min_x = min(A.x, B.x, C.x, D.x)
        max_x = max(A.x, B.x, C.x, D.x)
        min_y = min(A.y, B.y, C.y, D.y)
        max_y = max(A.y, B.y, C.y, D.y)

        if min_x <= P.x <= max_x and min_y <= P.y <= max_y:
            return True
        else:
            return False

    def zone_callback(self, msg):
        point = msg.point
        if msg.header.frame_id == "P1":
            self.P1 = (point.x, point.y)
        elif msg.header.frame_id == "P2":
            self.P2 = (point.x, point.y)
        elif msg.header.frame_id == "P3":
            self.P3 = (point.x, point.y)
        elif msg.header.frame_id == "P4":
            self.P4 = (point.x, point.y)

def main():
    rospy.init_node('zone_node')
    zone = Zone()
    rospy.Subscriber('/NavSat/fix', NavSatFix, zone.gps_callback)
    rospy.Subscriber('/Area/Point', PointStamped, zone.zone_callback)
    rospy.spin()

if __name__ == '__main__':
    main()

File: GPS/Zone.cpp
#include "zone.hpp"

       
        void Zone::gpsCallback(const sensor_msgs::NavSatFix::ConstPtr& msg) {
            // Callback appelée lors de la réception de messages GPS

            // Obtient les coordonnées GPS
            global_latitude = msg->latitude;
            global_longitude = msg->longitude;

            // Utilisez les coordonnées comme nécessaire (par exemple, les stocker dans des variables globales)
            ROS_INFO("Coordonnées GPS : Latitude = %f, Longitude = %f", global_latitude, global_longitude);
        }



        // Fonction pour calculer le produit vectoriel de deux vecteurs
        double Zone::crossProduct(Point A, Point B, Point C) {
            return (B.x - A.x) * (C.y - A.y) - (B.y - A.y) * (C.x - A.x);
        }

        // Fonction pour vérifier si un point est à l'intérieur d'un quadrilatère
        bool Zone::isInsideQuadrilateral(Point A, Point B, Point C, Point D, Point P) {
            // Calculer les produits vectoriels pour chaque côté du quadrilatère
            double crossABP = crossProduct(A, B, P);
            double crossBCP = crossProduct(B, C, P);
            double crossCDP = crossProduct(C, D, P);
            double crossDAP = crossProduct(D, A, P);

            // Vérifier si le point est à l'intérieur du quadrilatère
            if ((crossABP >= 0 && crossBCP >= 0 && crossCDP >= 0 && crossDAP >= 0) ||
                (crossABP <= 0 && crossBCP <= 0 && crossCDP <= 0 && crossDAP <= 0)) {
                return true;
            } else {
                return false;
            }
        }
    


void Zone::ZoneCallback(const geometry_msgs::PointStamped & msg){

    switch (Area.header.frame_id){

    case "P1":
        My_Zone.P1 = {Area.point.x, Area.point.y};
        break;
    case "P2":
        My_Zone.P2 = {Area.point.x, Area.point.y};
        break;
    case "P3":
        My_Zone.P3 = {Area.point.x, Area.point.y};
        break;
    case "P4":
        My_Zone.P4 = {Area.point.x, Area.point.y};
        break;
    }


}
int main() {
    double tab[4][3]
    
    // while (1){}
        // lire le topic /Area/Point (Les points de la zone)
        // lire le topic /NavSat/fix (Position GPS du robot)
        // Si data dans Area/Point  
            // cree nouvel objet Zone 
            // defninir P1,P2,P3
    My_Zone();

    return EXIT_SUCCES;
}
